### PR TITLE
Make the current flex facet require WTP 

### LINKED ISSUE #2034: NPE when deploying App Engine flex project not configured with Dynamic Web Module facet
Currently, you can add the AE flex facet to a plain Java project without adding the Dynamic Web Module facet. This seems to cause an NPE when trying to deploy such a project.

This is also distantly related to https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/2019#issuecomment-306873454.

### PR DESCRIPTION
**Pull request by chanseokoh**
Fixes #2034. As discussed in https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/2034#issuecomment-307829079, we will make the current flex facet require WTP.

### CODE REVIEW

File: plugins/com.google.cloud.tools.eclipse.appengine.facets.test/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineFlexFacetTest.java [hunk 1]
```diff
@@ -65,4 +69,53 @@ public void testFacetLabel() {
...
+  public void testInstallConstraints_notOkWithNoServlet() {
+    IConstraint constraint = AppEngineFlexFacet.FACET_VERSION.getConstraint();
+    IStatus result = constraint.check(Arrays.asList(JavaFacet.VERSION_1_8));
+    Assert.assertFalse(result.isOK());
+  }
+
+  @Test
+  public void testInstallConstraints_notOkWithJava7Servlet31() {
+    IConstraint constraint = AppEngineFlexFacet.FACET_VERSION.getConstraint();
+    IStatus result = constraint.check(Arrays.asList(JavaFacet.VERSION_1_7, WebFacetUtils.WEB_31));
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets.test/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineFlexFacetTest.java]: Why is this case invalid?

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/plugin.xml [hunk 2]
```diff
@@ -113,15 +113,33 @@
...
+      <label>%flexFacetName</label>
+      <description>%flexFacetDescription</description>
     </project-facet>
     <project-facet-version facet="com.google.cloud.tools.eclipse.appengine.facets.flex" version="1">
-       <constraint>
+      <constraint>
+        <and>
           <conflicts facet="com.google.cloud.tools.eclipse.appengine.facets.standard"/>
-        </constraint>
+          <requires group="com.google.cloud.tools.eclipse.appengine.facets.flex.constraints.servlet"/>
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/plugin.xml]: I'm regretting this approach for the Java 8 support.  Although groups are extensible, they don't allow specifying additional constraints: it's just "at least one from this group".

File: plugins/com.google.cloud.tools.eclipse.appengine.facets.test/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineFlexFacetTest.java (hunk 1 above)
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets.test/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineFlexFacetTest.java]: My understanding is that we don't have a Java 7 runtime: https://cloud.google.com/appengine/docs/flexible/java/dev-jetty9. The Java 8 / Jetty 9.3 Runtime for WARs, and Java 8 Runtime for JARs.

**Comment by chanseokoh** [comment]
For the Java 8 / Jetty 9.3 Runtime (provided for WARs), I think these are the right constraints? Of course, the flex env is really about running as a container, and your own custom runtime can be anything you can think of, but for Java WARs which we support by the current flex facet design, we use this particular runtime. We should soon think about how we would support the Java 8 Runtime (for JARs) or a custom runtime.

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/plugin.xml (hunk 2 above)
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/plugin.xml]: So, what do you think will be better? Using `<require facet=... version=...>` constraints directly without the indirect group constraints?

**Comment by elharo** [comment]
shouldn't an app compiled for Java 7 run on the Java 8 runtime?

**Comment by chanseokoh** [comment]
> shouldn't an app compiled for Java 7 run on the Java 8 runtime?

Ah, I get it.

File: plugins/com.google.cloud.tools.eclipse.appengine.facets.test/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineFlexFacetTest.java [hunk 3]
```diff
@@ -73,21 +73,21 @@ public void testFacetLabel() {
   @Test
   public void testInstallConstraints_okWithJava7Servlet25() {
     IConstraint constraint = AppEngineFlexFacet.FACET_VERSION.getConstraint();
-    IStatus result = constraint.check(Arrays.asList(JavaFacet.VERSION_1_8, WebFacetUtils.WEB_25));
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets.test/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineFlexFacetTest.java]: Oops, missed these :blush:

**Pull request merged**